InternalServerError from APNs

We are using Pushy in our production environment and it works great! Besides we have a staging environment where tests are performed prior to launching new versions. Staging environment is not always used and most of the time it is in idle state.
Today when we started to use staging, we got `InternalServerError` errors (stack trace given below). I guess it is related to long idle state of pushy, not sure though. We didn't have any choice but to restart the app.

In which cases APNs may return such an error? Should pushy be closing and reconnecting in such cases? Or is it a duty of our side to reconnect upon such an error?

Using pushy v0.9, java 1.8.0_102 (Oracle)

    2017-01-06 05:23:27,715  WARN [nioEventLoopGroup-2-4] c.r.p.a.ApnsClient APNs server reported an internal error when sending ... .
    2017-01-06 05:23:27,715 ERROR [nioEventLoopGroup-2-4] o.g.g.a.ApnsResponseListener Failed to send notification id=117_963041067
    java.util.concurrent.ExecutionException: com.relayrides.pushy.apns.ApnsServerException: {"reason":"InternalServerError"}
	    at io.netty.util.concurrent.AbstractFuture.get(AbstractFuture.java:54)
	    at org.grouvi.gpns.apns.ApnsResponseListener.operationComplete(ApnsResponseListener.java:59)
	    at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:514)
	    at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:507)
	    at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:486)
	    at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:427)
	    at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:129)
	    at com.relayrides.pushy.apns.ApnsClient.handleServerError(ApnsClient.java:965)
	    at com.relayrides.pushy.apns.ApnsClientHandler$ApnsClientHandlerFrameAdapter.onDataRead(ApnsClientHandler.java:173)
	    at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onDataRead(DefaultHttp2ConnectionDecoder.java:229)
	    at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readDataFrame(DefaultHttp2FrameReader.java:411)
	    at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:245)
	    at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:155)
	    at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:118)
	    at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:335)
	    at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:395)
	    at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351)
	    at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:266)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351)
	    at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1069)
	    at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:902)
	    at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351)
	    at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	    at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:651)
	    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:574)
	    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:488)
	    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:450)
	    at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873)
	    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
	    at java.lang.Thread.run(Thread.java:745)
    Caused by: com.relayrides.pushy.apns.ApnsServerException: {"reason":"InternalServerError"}
	    ... 37 common frames omitted

*Another observation*:

To get rid of `InternalServerError` exceptions we are restarting the app and when stopping we see plenty of error logs with `ClientNotConnectedException` (stack trace below). I am pretty sure that those exceptions are for old push notifications, not for those that were queued when jvm was terminating. According to logs, notifications were not accepted and neither rejected by APNS. So it seems listener threads hung up waiting for the future result.

I have suspicions that this may be related to the issue. Not sure though.

I have to mention that we use custom event loop group with 4 threads and all error logs were from the same thread, in stack trace below - *nioEventLoopGroup-2-2*

    2017-01-30 08:31:24,300 ERROR [nioEventLoopGroup-2-2] o.g.g.a.ApnsResponseListener Failed to send notification id=67087_1985304421
     java.util.concurrent.ExecutionException: com.relayrides.pushy.apns.ClientNotConnectedException: Client disconnected unexpectedly.
            at io.netty.util.concurrent.AbstractFuture.get(AbstractFuture.java:54)
            at org.grouvi.gpns.apns.ApnsResponseListener.operationComplete(ApnsResponseListener.java:59)
            at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:514)
            at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:507)
            at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:486)
            at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:427)
            at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:129)
            at com.relayrides.pushy.apns.ApnsClient$2$2.run(ApnsClient.java:495)
            at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
            at io.netty.util.concurrent.PromiseTask.run(PromiseTask.java:73)
            at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
            at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:418)
            at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:454)
            at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873)
            at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
            at java.lang.Thread.run(Thread.java:745)
    Caused by: com.relayrides.pushy.apns.ClientNotConnectedException: Client disconnected unexpectedly.
            ... 9 common frames omitted

--- a/pushy/src/main/java/com/turo/pushy/apns/ApnsClientHandler.java
+++ b/pushy/src/main/java/com/turo/pushy/apns/ApnsClientHandler.java
@@ -75,6 +75,8 @@ class ApnsClientHandler extends Http2ConnectionHandler implements Http2FrameList
     private static final IOException STREAM_CLOSED_BEFORE_REPLY_EXCEPTION =
             new IOException("Stream closed before a reply was received");
 
+    private static final ApnsServerException APNS_SERVER_EXCEPTION = new ApnsServerException();
+
     private static final Gson GSON = new GsonBuilder()
             .registerTypeAdapter(Date.class, new DateAsTimeSinceEpochTypeAdapter(TimeUnit.MILLISECONDS))
             .create();
@@ -287,18 +289,55 @@ class ApnsClientHandler extends Http2ConnectionHandler implements Http2FrameList
 
         if (endOfStream) {
             final Http2Stream stream = this.connection().stream(streamId);
-
-            final Http2Headers headers = stream.getProperty(this.responseHeadersPropertyKey);
-            final ApnsPushNotification pushNotification = stream.getProperty(this.pushNotificationPropertyKey);
-
-            final ErrorResponse errorResponse = GSON.fromJson(data.toString(StandardCharsets.UTF_8), ErrorResponse.class);
-
-            this.handleErrorResponse(context, streamId, headers, pushNotification, errorResponse);
+            this.handleEndOfStream(context, this.connection().stream(streamId), (Http2Headers) stream.getProperty(this.responseHeadersPropertyKey), data);
         } else {
             log.error("Gateway sent a DATA frame that was not the end of a stream.");
         }
 
         return bytesProcessed;
+    }
+
+    @Override
+    public void onHeadersRead(final ChannelHandlerContext context, final int streamId, final Http2Headers headers, final int streamDependency, final short weight, final boolean exclusive, final int padding, final boolean endOfStream) throws Http2Exception {
+        this.onHeadersRead(context, streamId, headers, padding, endOfStream);
+    }
+
+    @Override
+    public void onHeadersRead(final ChannelHandlerContext context, final int streamId, final Http2Headers headers, final int padding, final boolean endOfStream) throws Http2Exception {
+        log.trace("Received headers from APNs gateway on stream {}: {}", streamId, headers);
+        final Http2Stream stream = this.connection().stream(streamId);
+
+        if (endOfStream) {
+            this.handleEndOfStream(context, stream, headers, null);
+        } else {
+            stream.setProperty(this.responseHeadersPropertyKey, headers);
+        }
+    }
+
+    private void handleEndOfStream(final ChannelHandlerContext context, final Http2Stream stream, final Http2Headers headers, final ByteBuf data) {
+
+        final ApnsPushNotification pushNotification = stream.getProperty(this.pushNotificationPropertyKey);
+        final Promise<PushNotificationResponse<ApnsPushNotification>> responsePromise = stream.getProperty(this.responsePromisePropertyKey);
+
+        final HttpResponseStatus status = HttpResponseStatus.parseLine(headers.status());
+
+        if (HttpResponseStatus.OK.equals(status)) {
+            responsePromise.trySuccess(
+                    new SimplePushNotificationResponse<>(pushNotification, true, null, null));
+        } else {
+            if (HttpResponseStatus.INTERNAL_SERVER_ERROR.equals(status)) {
+                log.warn("APNs server reported an internal error when sending {}.", pushNotification);
+                responsePromise.tryFailure(APNS_SERVER_EXCEPTION);
+                context.channel().close();
+            } else {
+                if (data != null) {
+                    final ErrorResponse errorResponse = GSON.fromJson(data.toString(StandardCharsets.UTF_8), ErrorResponse.class);
+                    this.handleErrorResponse(context, stream.id(), headers, pushNotification, errorResponse);
+                } else {
+                    log.warn("Gateway sent an end-of-stream HEADERS frame for an unsuccessful notification.");
+                }
+            }
+        }
     }
 
     protected void handleErrorResponse(final ChannelHandlerContext context, final int streamId, final Http2Headers headers, final ApnsPushNotification pushNotification, final ErrorResponse errorResponse) {
@@ -313,39 +352,6 @@ class ApnsClientHandler extends Http2ConnectionHandler implements Http2FrameList
         } else {
             responsePromise.trySuccess(new SimplePushNotificationResponse<>(pushNotification,
                     HttpResponseStatus.OK.equals(status), errorResponse.getReason(), errorResponse.getTimestamp()));
-        }
-    }
-
-    @Override
-    public void onHeadersRead(final ChannelHandlerContext context, final int streamId, final Http2Headers headers, final int streamDependency, final short weight, final boolean exclusive, final int padding, final boolean endOfStream) throws Http2Exception {
-        this.onHeadersRead(context, streamId, headers, padding, endOfStream);
-    }
-
-    @Override
-    public void onHeadersRead(final ChannelHandlerContext context, final int streamId, final Http2Headers headers, final int padding, final boolean endOfStream) throws Http2Exception {
-        log.trace("Received headers from APNs gateway on stream {}: {}", streamId, headers);
-        final Http2Stream stream = this.connection().stream(streamId);
-
-        if (endOfStream) {
-            final HttpResponseStatus status = HttpResponseStatus.parseLine(headers.status());
-            final boolean success = HttpResponseStatus.OK.equals(status);
-
-            if (!success) {
-                log.warn("Gateway sent an end-of-stream HEADERS frame for an unsuccessful notification.");
-            }
-
-            final ApnsPushNotification pushNotification = stream.getProperty(this.pushNotificationPropertyKey);
-            final Promise<PushNotificationResponse<ApnsPushNotification>> responsePromise = stream.getProperty(this.responsePromisePropertyKey);
-
-            if (HttpResponseStatus.INTERNAL_SERVER_ERROR.equals(status)) {
-                log.warn("APNs server reported an internal error when sending {}.", pushNotification);
-                responsePromise.tryFailure(new ApnsServerException());
-            } else {
-                responsePromise.trySuccess(
-                        new SimplePushNotificationResponse<>(pushNotification, success, null, null));
-            }
-        } else {
-            stream.setProperty(this.responseHeadersPropertyKey, headers);
         }
     }
 

--- a/pushy/src/test/java/com/turo/pushy/apns/ApnsClientTest.java
+++ b/pushy/src/test/java/com/turo/pushy/apns/ApnsClientTest.java
@@ -511,11 +511,14 @@ public class ApnsClientTest {
         try {
             terribleTerribleServer.registerVerificationKey(this.verificationKey, DEFAULT_TOPIC);
 
+            final TestMetricsListener metricsListener = new TestMetricsListener();
+
             final ApnsClient unfortunateClient = new ApnsClientBuilder()
                     .setApnsServer(HOST, PORT)
                     .setSigningKey(this.signingKey)
                     .setTrustedServerCertificateChain(CA_CERTIFICATE)
                     .setEventLoopGroup(EVENT_LOOP_GROUP)
+                    .setMetricsListener(metricsListener)
                     .build();
 
             try {
@@ -530,6 +533,11 @@ public class ApnsClientTest {
                 assertTrue(future.isDone());
                 assertFalse(future.isSuccess());
                 assertTrue(future.cause() instanceof ApnsServerException);
+
+                unfortunateClient.sendNotification(pushNotification).await();
+
+                assertTrue("Connections should be replaced after receiving an InternalServerError.",
+                        metricsListener.getConnectionsRemoved().get() == 1);
             } finally {
                 unfortunateClient.close().await();
                 Thread.sleep(10);

